### ISSUE TITLE
Automatically follow user's OS-level dark/light preference

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#4009]
Over the last year or two it's become common for apps to have both a dark and a light color scheme, and let the user choose between them. Zulip has this.

Since 2019 this has become a thing at the OS level. And in recent months, I've seen more and more apps noticing my OS-level setting and following along. We should do the same, at least by default -- when the user goes into system-wide settings and says they want the device's screen to stay dark instead of bright, we shouldn't wait for an individual invitation.

This becomes especially significant now that it's an OS-level feature, in both Android and iOS, to have the light/dark mode switch back and forth every morning and evening. The only way that can work as a good experience is when apps all follow along automatically.

[RN v0.62 adds an API](https://reactnative.dev/blog/2020/03/26/version-0.62#new-dark-mode-features) for this. ~~So I think it makes sense to hold off on trying to implement this behavior until we've done that upgrade, aka #3782.~~ Done!

**Comment by chrisbobbe** [comment]
As part of this, let's deprecate the 'default' name, like so:

```diff
- export type ThemeName = 'default' | 'night'
+ export type ThemeName = 'light' | 'night'
```

As I said over at https://github.com/zulip/zulip-mobile/pull/4046#discussion_r414032884:

> Hmm. Taking a look at [Apple's HIG document on Dark Mode](https://developer.apple.com/design/human-interface-guidelines/ios/visual-design/dark-mode/), though, it makes it sound like "default" semantics shouldn't apply to either theme at the application level, but rather at the device level:
> 
> > In Settings, people can choose Dark Mode as their default interface style and schedule automatic changes between the appearance modes. Because people make these choices at a systemwide level, they generally expect all apps to respect their preferences.
> 
> Perhaps we might actually deprecate the "default" name, and use "night" or "light" everywhere (with our `ThemeName` type being `'light' | 'night'` instead of `'default' | 'night'` as it is currently), at or on the path to #4009.

and Greg [agreed](https://github.com/zulip/zulip-mobile/pull/4046#discussion_r414886033).

**Comment by gnprice** [comment]
It'd also be good to switch to calling it "dark mode" instead of "night mode" or "night theme" -- as "dark" is what both Google and Apple call it, and also is more logical because there's no inherent reason to only use it at night.

That, or at least the UI side of it, should happen in coordination with the webapp, though, so the same terminology is used across platforms.

**Comment by gnprice** [comment]
> It'd also be good to switch to calling it "dark mode" instead of "night mode" or "night theme" -- as "dark" is what both Google and Apple call it, and also is more logical because there's no inherent reason to only use it at night.

(To be more precise on the Google side: it's "dark theme" in the UI and [user-facing docs](https://support.google.com/accessibility/android/answer/6151800), and they follow that in the [Material docs](https://material.io/develop/android/theming/dark) and [guide-style developer docs](https://developer.android.com/guide/topics/ui/look-and-feel/darktheme). It's "night mode" [in the API](https://developer.android.com/reference/android/app/UiModeManager#setNightMode(int)), reflecting API choices made for API 8 aka Android 2.2 Froyo, in 2010.)

**Comment by Gautam-Arora24** [comment]
I would like to work on this issue @chrisbobbe and @gnprice 😃

**Comment by chrisbobbe** [comment]
When in "follow the OS" mode, we'll want to make sure to listen for changes in the OS-level setting and apply them automatically, right away.

For that, I could imagine an approach that uses [`Appearance.addChangeListener()`](https://reactnative.dev/docs/0.63/appearance#addchangelistener), with that listener logic probably living in src/boot/AppEventHandlers.js.

Or, one that uses the [`useColorScheme` hook](https://reactnative.dev/docs/0.63/usecolorscheme).

I think I slightly prefer using the `useColorScheme` hook, as I've been keeping an eye on the growth of `AppEventHandlers` and wondering if there's a better way we could handle the things it handles.

`state.settings.theme` is currently what the UI logic reads to decide whether to show light or dark components (see src/boot/ThemeProvider.js). It has two possible values, currently: one for light mode, one for dark mode.

**Comment by chrisbobbe** [comment]
For explicitness, I think the task is to change the current "Night mode" switch on the settings screen, so that it's called "Theme" (or similar) and lets the user choose among these three options (naming TBD):
- Light
- Dark
- Follow OS

And to rewire our Redux logic for this setting, and also use that bit of React Native API Greg mentioned, so the setting works as expected.

**Comment by BigHeadCreations** [comment]
Hey there @chrisbobbe and @gnprice, I'll take a shot at this Issue.

**Comment by chrisbobbe** [comment]
Great, thanks! I see you've started a [discussion on CZO](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Dark.20mode.20preference/near/1451399); thanks! 🙂

**Comment by alya** [comment]
Does resolving this issue also solve #5169?

**Comment by gnprice** [comment]
Yep, it does.

**Comment by gnprice** [comment]
This is still a feature we want, but we've just discovered -- see https://github.com/zulip/zulip-mobile/pull/5527#issuecomment-1302588873 -- that our intended approach for solving it won't work (due to a bug in React Native) and we'll need something a bit more complex.

Because this issue thread has gotten long, I just filed a fresh issue laying out the new approach: #5533. Closing this one in favor of that.
